### ISSUE TITLE
Port all patchers to Skyrim

### ISSUE THREAD

**Issue by Sharlikran** [wrye-bash/wrye-bash#151]
I am currently working on and testing changes to Skyrim's patchers.  All changes for Skyrim and Oblivion should be made to 151-skyrim-patchers and wrye-bash contributors should comment on changes in this issue.  151-skyrim-patchers will be rebased off of dev each time Utumno merges changes.

Branch: https://github.com/Sharlikran/wrye-bash/tree/dev-skyrim-tes4-update

Tasklist:
- [x] Decide whether or not to fully update LAND since it normally consumes a lot of time to process
- [x] Update complex records from TES5Edit to simple records when applicable
- [x] Update remaining records LAND, PERK, PACK, RACE, SCEN -> #480
- [x] Update existing 305 Patcher support for mergable records
- [x] Update remaining Patcher support for mergable records
- [x] Add support for Fragment Scripts for 'INFO' in VMAD

bosh.py Patchers to update:
- [x] Alias Mod Names
- [x] Contents Checker
- [x] Import Actors
- [x] Import Actors: AI Packages
- [x] Import Actors: Death Items
- [x] Import Actors: Factions
- [x] Import Actors: Spells
- [x] Import Cells
- [x] Import Destructible
- [x] Import Effect Stats
- [x] Import Enchantment Stats
- [x] Import Graphics
- [x] Import Inventory
- [x] Import Keywords
- [x] Import Names
- [x] Import Object Bounds
- [x] Import Outfits
- [x] Import Relations
- [x] Import Sounds
- [x] Import Spell Stats
- [x] Import Stats
- [x] Import Text
- [x] Leveled Lists
- [x] Merge Patches
- [x] Tweak Actors
- [x] Tweak Assorted
- [ ] Tweak Names
- [x] Tweak Settings
- [x] Race Records

Needs to be clarified or investigated:
- [x] DESC subrecords do not appear in the Bash Patch when merging BOOK
  - Had dupe attribute 'description'
- [x] In `bosh.py: class PatchFile` [what does this do](https://github.com/Sharlikran/wrye-bash/commit/d606140ab342a1718eb50f09bc2bbeb4ece37965#diff-1f37627f4d9ab352fab2c141c307ea09R10507)?  Is this Skyrim or Oblivion specific? - Infernio: That is Oblivion-specific. SCPT doesn't even exist in Skyrim (due to Papyrus). Probably needs to be expanded for FO3 and FNV as well though.
- [x] FormID Lists: new issue -> #497
- [x] Import Scripts: Can this be used for VMAD without destroying VMAD's functionality? - in theory yes, but it's not a good idea to touch VMAD since it interacts directly with the external script files

Note: most of Skyrims records are updated excluding those mentioned.

**Comment by Sharlikran** [comment]
Commit b583e1202f1f0b587ff8856dfdefd809cab84a81 will no longer rebase without resolving changes.  I no longer wish to try and follow the changes.  I'll leave this open as it is something that needs to be done but I won't be working on it.

**Comment by Utumno** [comment]
Which branch you want rebased ? I can resolve the conflicts probably ;)

**Comment by Utumno** [comment]
Done: https://github.com/wrye-bash/wrye-bash/tree/dev-sharl

The conflicts were trivial happily (for me who knew what I have changed :))

Is some of it ripe enough to be merged ?

**Comment by Sharlikran** [comment]
You rebased my feature branch which has some of the Skyrim and Fallout patchers active while attempting to keep Oblivion functionality in tact.  Naturally the CBash side of things remains untouched. However, as I mentioned I'm not going to follow the refactoring anymore.

It has been tested a lot but there are issues as noted in the OP of the dev thread under Known Issues.  Nobody would use it for Fallout since it doesn't patch enough of the same things that Valda's version patches.

I wouldn't merge this into dev because once you do then after people have downloaded it and start making patches for Skyrim they will begin to point out all the flaws in the patchers.  Keywords and VMAD records won't be carried over into the Bash Patch correctly and all kinds of stuff.

The intent was for me to work with Valda and get most of the issues resolved.  However, he quit talking to me a long time ago.  Since there is still refactoring to do feel free to do that and you are doing great work.

Until the refactoring can stop 100% and all focus is spent on patchers then like WrinklyNinja, I am uninterested in working with the project.

**Comment by Utumno** [comment]
A shame

Please pull this branch (dev-sharl) in your clone for future reference - let me know when done so I can delete it from here.

**Comment by Utumno** [comment]
Plus - when I say merge I mean what is ready to be merged from _the work on skyrim records_. Are there pathers skyrim pathers that work correctly ? Those could be enabled too.

**Comment by Utumno** [comment]
Finally - update the OPs here and in #150 with the links to the branches you mention - I rebased on feature-sharlikran-valda-fallout cause that is what was closer to what you have in the OP (broken link)

Thanks

**Comment by Sharlikran** [comment]
Branch used in this pull request:

[dev-skyrim-patcher-update](https://github.com/Sharlikran/wrye-bash/tree/dev-skyrim-patcher-update)
```
dev
   \ dev-skyrim-patcher-update
```

You don't have to but if you have time This pull request is based off of dev and only has the changes found in the dev-skyrim-patcher-update branch.

**Comment by Utumno** [comment]
Ok - here you are: https://github.com/wrye-bash/wrye-bash/tree/dev-sharl-skyrim.
Please pull it to your tree and force update the pull request - let me know when done cause I have to delete it from here. I will keep updating dev but no conflicts will occur.

**Comment by Utumno** [comment]
Hey - deleted the dev-sharl branch as I had omissed a line:

     +        progress.setFull(len(self.srcs))

It only differs with the fallout branch in the last few commits - just cherry-pick them (gitk, right click) - actually only the _added build release batch file_
 should be needed)

**Comment by Sharlikran** [comment]
At first I wasn't able to Rebase mostly because I couldn't resolve the conflicts very easily.  I decided to try and take the time and make a better attempt at resolving the conflicts myself by comparing the code from dev and the code I am using for the test version I last posted in the Experimental Bethesda thread.  That code was based off of changes as of commit 57891d5e66a0b4737983ac4a44f689e8aa1001ae.

While comparing the changes I noticed that you made a major change that i would have to undo.

```Python
#--Patch Phase ------------------------------------------------------------
    def initPatchFile(self,patchFile,loadMods):
        super(CellImporter, self).initPatchFile(patchFile,loadMods)
        self.cellData = {}
        # TODO: docs: recAttrs vs tag_attrs - extra in PBash:
        # 'unused1','unused2','unused3'
        self.recAttrs = {
            u'C.Climate': ('climate',),
            u'C.Music': ('music',),
            u'C.Name': ('full',),
            u'C.Owner': ('ownership',),
            u'C.Water': ('water','waterHeight'),
            u'C.Light': ('ambientRed','ambientGreen','ambientBlue','unused1',
                        'directionalRed','directionalGreen','directionalBlue','unused2',
                        'fogRed','fogGreen','fogBlue','unused3',
                        'fogNear','fogFar','directionalXY','directionalZ',
                        'directionalFade','fogClip'),
            u'C.RecordFlags': ('flags1',), # Yes seems funky but thats the way it is
            }
        self.recFlags = {
            u'C.Climate': 'behaveLikeExterior',
            u'C.Music': '',
            u'C.Name': '',
            u'C.Owner': 'publicPlace',
            u'C.Water': 'hasWater',
            u'C.Light': '',
            u'C.RecordFlags': '',
            }
```
That code can't be there like that because each game has different attributes since each game mode has a different record structure.

```Python
#--Patch Phase ------------------------------------------------------------
    def initPatchFile(self,patchFile,loadMods):
        super(CellImporter, self).initPatchFile(patchFile,loadMods)
        self.cellData = {}
        # TODO: docs: recAttrs vs tag_attrs - extra in PBash:
        # 'unused1','unused2','unused3'
        self.recAttrs = game.cellRecAttrs
        self.recFlags = game.cellRecFlags```
The routine has to look like that.

**Comment by Sharlikran** [comment]
```Python
        # TODO: docs: recAttrs vs tag_attrs - extra in PBash:
        # 'unused1','unused2','unused3'
```
Also those attributes have to be there or the entire record won't be compared and patched correctly.  When Wrye Bash compares records for changes it has to look at all the attributes even things marked 'unused1','unused2','unused3' or 'unknown1','unknown2','unknown3'.  Otherwise a record that has no changes except changes made by the GECK, CS (Oblivion), or CK (Skyrim) would be considered altered.  

Those unused or unknown fields are undecoded data that can't be altered by the official Bethesda editors.  Since it can't be altered by the Official editors we assume it's something Bethesda uses internally and we can't determine what the fields are for.  However, they have to be evaluated as part of the record.  This is similar to how TES5Edit or TES4Edit determines an ITM.

**Comment by Sharlikran** [comment]
So again don't merge this into dev for 306 because it won't work right.  I have to wait until you are done refactoring and redo all my branches from scratch.  It won't matter if I use GitK or not because that's not the issue.  It's also not that I don't know Git well enough yet.  

It's simply that for the code to be separate for Skyrim and Oblivion there has to be constants in the game/oblivion and the game/skyrim directory for each game mode.  The same goes for the Fallout versions.

**Comment by Sharlikran** [comment]
Also I'm angry or disparaging your work.  You are doing great.  It's just a pain for me to follow since i don't know python hardly at all.

**Comment by Sharlikran** [comment]
delete dev-sharl-skyrim

**Comment by Utumno** [comment]
This comment was mine cause I was wondering on the meaning of these _unknown_ flags - compare also the differences between C and P patches `docs: recAttrs vs tag_attrs` - thanks for the clarifications - adding comments in the code for this would be great ( `TODO: docs: ....`)

The 

    self.recAttrs = game.cellRecAttrs
    self.recFlags = game.cellRecFlags

should end up in `<game>/__init__.py` for each game as you say - maybe introduce another file like `<game>/patcher.py` to have all these constants there ? I have been thinking this for a long time actually...

**Comment by Utumno** [comment]
@Sharlikran - I have some good news: @jfpelland is interested in helping with this. Is the branch at https://github.com/wrye-bash/wrye-bash/compare/dev...Sharlikran:dev-skyrim-patcher-update the latest iteration ?

If yes it would be nice if you squashed together the taglist commits (that should be probably committed in dev for release) and the loot commits (what are those about ?) - and move them at the top of the branch - except if the patcher changes depend on those, in which case they should be moved on the bottom of the branch.

Am I right to believe that you are nowadays rather comfortable with git rebase ? ;)

**Comment by Sharlikran** [comment]
yes https://github.com/Sharlikran/wrye-bash/commits/dev-skyrim-patcher-update is the current Skyrim branch.  He is already part of my collaborators branch.  I don't want to squash the taglist commits.  I have one for Oblivion and one for Skyrim because there are over 250ish and 350ish changes.  If I put them together you don't see what the changes are.  I had to rebuild that branch recently so nothing needs to be squished yet.

I'll talk to him and see what he is willing to do.  If he wants to know what I am looking for then I already know what needs worked on.

Also your instructions to use Notepad doesn't work for me.  I tried setting it up a few weeks ago.  I don't have that file you mention and even if I make one with the syntax you mentioned Git does not tell Notepad++ to open and display the commits as you explain.  I tired Git Extensions but it wants to append the commit and a bunch of things I don't understand yet.  I still have to figure out how to Squash commits an easier way.

**Comment by Sharlikran** [comment]
jfpelland, what did you want to work on?  If you are just wondering what I want done then I need MreRace finished for starters.  Then next after that I'd like to get the Faction Patcher altered.  As it is now it will work for Fallout and Oblivion.  However, Skyrim uses an entirely different setup.

**Comment by Utumno** [comment]
> I don't want to squash the taglist commits. I have one for Oblivion and one for Skyrim because there are over 250ish and 350ish changes. 

Perfectly fine - I remembered an older iteration where there were some 4-5 commits :)

Notepad ++: yes you do not have the file you have to write it - doesn't work ? Hmmm

Maybe forget my instructions and just do a google search - maybe there are more up to date/easier to follow instructions

Thanks for your continued support

**Comment by Utumno** [comment]
@Sharlikran: new and final branch on skyrim patchers: https://github.com/wrye-bash/wrye-bash/tree/151-skyrim-patchers

Please pull this in your fork and replace dev-skyrim-patcher-update. Also update your fork to be in sync with upstream (pull dev and rebase your main branch on `151-skyrim-patchers` and delete the now obsolete sharlikran-skyrim-rebase-on-dev - its content is in 151-skyrim-patchers). Keep loot separated as you have it but also rebase on 151-skyrim-patchers.

I will be adding commits in this branch, squashing warnings and moving patcher specific stuff from `game/<game>/__init__.py` to ~~`game/<game>/patcher/__init__.py`~~ `game/<game>/constants.py`. I will post comments on the commits for clarifications. I will also rewrite history, rebasing on dev and possibly squashing fixup commits.

Note I added the readme additions from your github.io repository

It would be better if you keep updating your fork as I rewrite 151-skyrim-patchers but we will be basing our discussion on the upstream branch (here). Once we are confident enough we will merge to dev and have a 307 beta out. Meanwhile it would be nice if you updated your EXP versions for the forums - just be sure to refer to which HEAD of 151-skyrim-patchers you base them.

@wrye-bash/active, @wrye-bash/bashers: if you use the python version pull this branch (or download the zip) and test. Be sure to redownload whenever this branch is updated.

Test also that oblivion patchers work - there was some shuffling going on

@jfpelland: I need to take over for the engineering part (package layout, warnings etc) but I still assign to you for the patchers part. Please pull and keep an eye open for changes.

**Comment by Sharlikran** [comment]
I'll work on that over the next couple of days. However, something has come to mind after my bout with Fallout 4 decoding when I don't even have a copy of the game yet so I don't have a Fallout 4.esm to work with. In my eagerness to get started with FO4Edit I'm putting the cart before the horse.  To do things the way you outlined in previous comments you want the patcher and the records that go together for each patcher. However, all of the records have to be completed and Merge Patches needs to be finished first because it's the foundation for everything.

![firstpatcher](https://cloud.githubusercontent.com/assets/1912426/11110450/d33a8a88-88b0-11e5-8b92-11e3d1be0c3e.jpg)

However to do that the remaining records need to function properly. Some records are a little to complex to merge like WRLD, and CELL so they won't be merged but the record has to be properly defined. 

As mentioned above I have been unable to update the following records: LAND, PERK, PACK, RACE, SCEN.  

The reason I can't do those is because they have sets of records which means duplicate signatures.  I can't simply say use NAM0 here and then NAM0 again in another location.  Wrye Bash won't accept that with it's current API. What has to happen is more like how RACE is defined for Fallout New Vegas.  Hence why I refer to Fallout so much when you don't want to confuse it with Skyrim and Oblivion.  However, that's not why I mention it.  I mention it because the record definitions are the foundation for what needs to happen in Skyrim and eventually for Fallout 4.

RACE is probably the best to get your feet wet so take a look at how [MreRace](https://github.com/Sharlikran/wrye-bash/blob/feature-sharlikran-valda-fallout/Mopy/bash/game/falloutnv/records.py#L3839) is structured.  I need to duplicate that for Skyrim.

I know you are busy with a lot of things so no rush.  However, this is where we should start.

**Comment by Arthmoor** [comment]
I have the game, so if someone wants to have me send them a data dump, I can, as long as you tell me how to do that :P 

Not comfortable with the idea of sending files to people who don't have the game though.

**Comment by Sharlikran** [comment]
@Arthmoor no worries, there isn't a data dump that will really help for FO4Edit.  Hlp and Zilav have things covered for now.  Feel free to talk about things in my experimental thread or on AFK.  I appreciate the offer but this is solely about getting Skyrim's Patchers working.

I'm just saying that to get Skyrim going in the right direction we need to use other examples from other Games to see how other authors contributing to this project defined those records, where we should start, and why.

**Comment by Utumno** [comment]
Thanks Sharl - ok I see now. I have not ever looked carefully at that code but when I do I will start from those records (MreRace ). And I do not need the records merged separately with their patchers finally - the branch is well structured. Just people test it and me finishing up my commits and good to go.

Have a look at my (WIP) commits to see where we're going (moving stuff to constants, has to be repeated for fallout).

**Comment by Utumno** [comment]
Rebased on dev and cleaned up a bit: https://github.com/wrye-bash/wrye-bash/compare/151-skyrim-patchers

Posted comments/head ups on the commits for you.

When you get round to it hash up a 307.04 - thanks for keeping up the good work :)

It is important that latest dev is tested - plus game package layout freezes (I solved a couple of very annoying issues here) which means we will soon merge this (and have an official 307 beta out)

**Comment by Sharlikran** [comment]
I have rebased dev-skyrim-patcher-update off of 151-skyrim-patchers.  That way you can cherry-pick/amend any commits any way you want.  I have some changes there now from what you asked about.  I'd leave it because I am reviewing other Skyrim related changes.  I'll let you know when I am done for today so you can cherry-pick them into 151-skyrim-patchers once you have time.

**Comment by Sharlikran** [comment]
I made some comments on the commits but I forgot to @Utumno you for them.  If you see them then I'm looking for clarification or syntax help.

**Comment by Utumno** [comment]
Replied to you there ;)

**Comment by Sharlikran** [comment]
> do as you please just we end up with one branch and if we edit this we notify each other - deal ?

Deal.  I can always look at the pull request if I need to for your instructions.

I'll let you know if I change 151-skyrim-patchers here as you requested.  I will continue to try and work with you in a manner you feel comfortable with and causes you the least amount of headaches.  It may be a lot of communication initially but once I am on the same page as you I think things will calm down and you can focus on what you need to.  Thanks for all the feedback while we work toward some kind of unified workflow.

**Comment by Utumno** [comment]
:+1:

So force push in this branch (151-skyrim-patchers from now on) your changes (the final version of your ex pull with the squashed commits) and let me know so I pull it.

Keep in mind I will keep pushing on dev so you rebase often and again let me know here. If making changes to commits let me know in the commit's message (easier to follow). It's your branch basically - except if I notify you of some change/merge I want to make.

Thanks Sharl

**Comment by Utumno** [comment]
Ah btw - update the OP with branch info and possibly what's done (check boxes)

**Comment by Sharlikran** [comment]
There are reports now on my recently rebased code where Wrye Bash [assigns Bash Tags](http://forums.bethsoft.com/topic/1534516-wipz-wrye-bashexp-skyrim-patcher-update-thread-3/?p=24869090) to files that don't have any Bash tags.

I am having them test with 151-skyrim-patchers : f9395fc7009bd211a30e07e4e4aa5fe96aad5721 to see if it still happens.  If it only happens in my rebased version that adds FO3, FNV support and minor changes to FO4 then I will probably have to quit rebasing the changes because as mentioned before regardless of how often I rebase the files I can't keep up with the changes.  

Even little things like changing ListPatcher to _ListPatcherPanel are not things I will understand.  There could be any number of regressions for an error like this to be occurring.  

Now I could be sharing my feelings a little prematurely because it could also be new routines that don't update old .dat files correctly from previous versions.  No way to know yet but, if, my rebasing caused this I'm gonna have to stop doing it because I don't know Python.  It would take to long and I can't instantly learn it and be at your level.

**Comment by Sharlikran** [comment]
So it is a new DLL I will have to track down the actual cause in the loot32.dll.  I am so relieved it's not something else.

**Comment by Utumno** [comment]
Yep - don't panick is advice number one here

I can understand your mystification with python (hey, took *me *a couple
years and I was already a programmer) but first I am stilll around and
secondly rebasing often you keep an eye on dev and see what changes - and I
believe you do understand what commits could affect what (more or less).

Now that refactoring hit bosh, soon every thing will be in its own module -
and Bash is no more this spagheti stuff where changing anything could
affect anything else.

So when I saw the reports and *having an eye in your changes * it was
immediatelly obvious to me that it would be the loot dll - that's what
produces the tags, nothing else (apart from manual assigning them). So if
you keep a close eye on dev you will be able to easily track down broken
stuff - _ListPatcherPanel was a recent change after all ;)

**Comment by Sharlikran** [comment]
I try but I feel there is some underlying reason I get discouraged easily.  Some people are OCD clinically and simply can't help doing certain things.  I think I just have a hard time dealing with frustration.  I know everyone does but for me I think it's just a bit more uncontrollable in certain situations.  

I know you are there and you are always helpful but in the end it's just overwhelming for me.  I found the cause and once I have time to reproduce the issue I can report it to WrinklyNinja.  Thanks for keeping an eye on things.

**Comment by Sharlikran** [comment]
As requested.  I have made changes to 151-skyrim-patchers, please pull in the changes.  Also because you changed more then what you mentioned in a4fa9f3a5d5aeef280bf30f42d96a0468e300d52 I had to append two commits after Updated MelCoed added MelNpcCoed.

**Comment by Utumno** [comment]
Well done will have a look ASAP - rebased on https://github.com/wrye-bash/wrye-bash/commit/8d189f8af381c9b56808261b8d148b7306ccaa9c - good job :+1:

**Comment by Sharlikran** [comment]
I have made changes to 151-skyrim-patchers, please pull in the changes.  I won't be able to do more today or tomorrow but please let me know if there is anything else and I can get to it as soon as I am able.

**Comment by Utumno** [comment]
Great - I pushed a branch here: https://github.com/wrye-bash/wrye-bash/compare/151-skyrim-patchers-records.py

I want you to go through the commits and see if you have anything to comment on. Also (because merges in records proved tricky - remember 306 dev corrupting mods on editing the description) make sure that all records edits are safe.

I added some commits I squash shadowing builtin warnings - especially type is very annoying as is used all over - including BAIN. Have a look. will comment on the commits that I need you to pay attention to ASAP.

**Comment by Utumno** [comment]
Aaand pushed 151-skyrim-patchers rebased on https://github.com/wrye-bash/wrye-bash/compare/151-skyrim-patchers-records.py. Pull in and release a beta I'd say so we get some final feedback

**Comment by Utumno** [comment]
Alright commented on some commits. I will be very busy till sunday but I am looking forward to your response. In short add me comments in the commits (especially those I commented on) - you can leave some of them as simply "Updated XXX" if you think it's trivial. Then rebase on your exp branches and pack a release with this one so we get some feedback. I intend to merge the https://github.com/wrye-bash/wrye-bash/compare/151-skyrim-patchers-records.py except probably the last commits (mine). My experience  shows that 99% of regressions are introduced by last minute commits :D

Once this branch is merged we go on with the rest. ideally we are done in a week I can find some little time ;)

**Comment by Sharlikran** [comment]
It would seem dumb to me to put things like, "set it to baseName in 2012 and later changed it to name because I didn't know the patcher didn't want name in 2012."  To me a doc string defines why a routine needs to be a certain way like, "The second Integer for MelNpcCoed is a FormID and thus must be mapped.  This only applies to MelNpc.  In all cases of MelCoed the second Integer is a faction Rank and should be treated as an integer not a FormID."  That is an important note and benefits the project in an intelligent way.

If I comment on all the things you asked about I'd just be putting "just because that's what I did cuz I didn't know what I was doing in 2012 and I can't squash it now because skyrim.py doesn't exist."  I made the update because the patcher wants the name attribute not baseName.  If I had known how to squash commits I would have done so when the file was still skyrim.py prior to this commit : 682bf130b36866ce26983c204eef1ba4288cac53

There will be no added benefit for me to comment on any of your requests because it won't benefit the project or the code in any intelligent way.

**Comment by Sharlikran** [comment]
I basically made rough drafts of skyrim.py on sourceforge in [Branches/Wrye Bash](http://sourceforge.net/p/oblivionworks/code/HEAD/tree/Branches/Wrye%20Bash/).  There is a good chance some things I defined back in 2009.  That's how long I have had skyrim.py defined and all I have ever done is just refine the definitions as I go when I learn something new.

**Comment by Utumno** [comment]
> "The second Integer for MelNpcCoed is a FormID and thus must be mapped. This only applies to MelNpc. In all cases of MelCoed the second Integer is a faction Rank and should be treated as an integer not a FormID." 

That's what I want ;)

Anywhere you think it's important

**Comment by Sharlikran** [comment]
I can't describe how Form List patchers work concisely.  They are similar to Leveled list patchers and I believe Valda used the Leveled list patcher as a model for the patcher.  You would have to ask @valda or @lojack5.  

There are probably TES5Edit scripts that patch the Form Lists using Elminster's Merge Patch logic without causing issues.  In the event that a list can't be sorted users need to read the authors description page or ask if the scripting relies on the order of references in the Form List.  There will always be situations where certain patchers need specific bash tags in order to avoid placing records in the Bash Patch.  Wrye Bash has always patched records that require some kind of mitigation, that's what it was designed for.

**Comment by Arthmoor** [comment]
As I mentioned in the other thread, Skyrim is the only game that cares about the order of formlists. The vanilla game has plenty of scripts that rely on things being in the exact order they were stored. So I'm kind of thinking it's too risky to allow for any form of mitigation there.

Yes, this is what Wrye Bash was made for, but at the same time, it's expected to do so reliably. We can't guarantee reliability with formlists on Skyrim.

**Comment by Sharlikran** [comment]
151-skyrim-patchers rebased.

**Comment by Utumno** [comment]
You ninja'd me :)

Fetching...

**Comment by Utumno** [comment]
Alright fetched and compared - force pushing my rebase _which also contains edits in fallout4 files_ which will clash with your fallout4 branch (you also missed in your rebase some type -> rec_type in ob/skyrim init py). Compare your rebase with force pushed 151-skyrim-patchers then rebase your branches and let me know if you need help resolving conflicts. Ideally pack an EXP release.

Eager to merge the rest

**Comment by Utumno** [comment]
In case you are confused the right way to pull an upstream rebased branch is:

- fire up the tree
- issue git fetch upstream 151-skyrim-patchers
- refresh the tre, checkout local branches, compare etc
- then force reset your local 151-skyrim-patchers branch to upstream fetched (most easily done in the tree)
- rebase the rest (or rather the topmost) of your branches on 151-skyrim-patchers

but I have a feeling that you know all this by now.

May the tree be with you

**Comment by Sharlikran** [comment]
Actually please help.  Would you please review and fix any regression or things I overlooked?  I did my best but, rebasing that last batch of changes was a bit outside of my scope of understanding.  I could barley follow the changes.

As for an EXP release I'll wait until you come to some kind of stopping point after you finish merging the rest of the changes from 151-skyrim-patchers.  Not many people use the exp release as media fire tracks the downloads.  I'd say between 20 and 70 people have downloaded that version but Mediafire doesn't track unique downloads.  Some people download both for archival purposes.  Also, at the moment the links to my EXP version on the Bethesda forums are for Fallout 4 as well.

> which also contains edits in fallout4 files which will clash with your fallout4

If I can't rebase all my branches easily because of the changes I'll probably stop working on the project.  My [Bethesda post](http://forums.bethsoft.com/topic/1534516-wipz-wrye-bashexp-skyrim-patcher-update-thread-3/?p=24898690) explains things pretty well.

**Comment by Sharlikran** [comment]
After you fix up 151-skyrim-patchers please let me know.  I'll try rebasing and see how it goes once you are finished.

**Comment by Utumno** [comment]
The branch is fixed what do you mean ? It is rebased on dev and contains some changes in fallout4 files that might clash with https://github.com/Sharlikran/wrye-bash/commits/268-fallout4-support.
If you want me to rebase I can do that - no need to link me to posts - just link me to the branch you want rebased on your fork.

**Comment by Sharlikran** [comment]
Wait then I misunderstood, so you are saying that you already repaired 151-skyrim-patchers with anything I might have missed or messed up?

**Comment by Utumno** [comment]
> force pushing my rebase which also contains edits in fallout4 files

;)

**Comment by Utumno** [comment]
Alright pushed your https://github.com/Sharlikran/wrye-bash/commit/ce6014cc28eaaf068684fc94d526f5f06b9bca15 rebased on dev and added some fixup/wrapping/code style commits I intend to squash with previous commits as noted in the messages. You try rebasing yourself and compare your rebase with https://github.com/wrye-bash/wrye-bash/commit/8abb6b02a4623a4b58a88e0023f74bd7a5301d38. Will leave couple comments for you there

**Comment by Utumno** [comment]
Alright pushed your https://github.com/Sharlikran/wrye-bash/commit/ce6014cc28eaaf068684fc94d526f5f06b9bca15 rebased on dev and added some fixup/wrapping/code style commits I intend to squash with previous commits as noted in the messages. 

https://github.com/wrye-bash/wrye-bash/compare/feature-sharlikran-valda-fallout

You try rebasing yourself and compare your rebase with https://github.com/wrye-bash/wrye-bash/commit/8abb6b02a4623a4b58a88e0023f74bd7a5301d38. Will leave couple comments for you there

Don't worry about authors will change that before final merge

**Comment by Utumno** [comment]
And new rebased branches over dev - have a look. https://github.com/wrye-bash/wrye-bash/compare/151-skyrim-patchers is ready to merge - I only need to revisit the cosaves commit (and change ownership in one commit). I intend to merge ASAP !

https://github.com/wrye-bash/wrye-bash/compare/feature-sharlikran-valda-fallout is your branch at https://github.com/Sharlikran/wrye-bash/commit/ce6014cc28eaaf068684fc94d526f5f06b9bca15 rebased on current 151-skyrim-patchers

**Comment by Sharlikran** [comment]
151-skyrim-patchers has been changed/rebased.

**Comment by Utumno** [comment]
You added three commits in here - the cosaves one I had left you comments please reply to me there. The others are safe to merge ? See #282 especially for https://github.com/wrye-bash/wrye-bash/commit/699aa254ad649039e333488aaaa0cd0ea4282b56

**Comment by Sharlikran** [comment]
151-skyrim-patchers has been changed/rebased because I didn't rebase it off of dev by mistake.

**Comment by Utumno** [comment]
Alright thanks Sharl makes sense. Now I am at this moment making a final rebase over dev rewording my commits etc. Will leave you comments on commits

**Comment by Sharlikran** [comment]
> Will leave you comments on commits

I probably won't have more time to work on anything Wrye related or comment on anything today or in the near future for the same reasons previously stated.  I basically just sat down a mustered up enough energy to look at things after the user report from the other day that I linked in #282.

**Comment by Sharlikran** [comment]
I tried to rebase my branches to see what https://github.com/Sharlikran/wrye-bash/tree/feature-sharlikran-valda-fallout would look like.  However, now that you pulled the old version to wrye-bash/feature-sharlikran-valda-fallout I don't know which is more correct.

**Comment by Utumno** [comment]
I started a final rebase yesterday but then the record hell came on me (frankly I felt desperate looking at that code, frustration is too little to describe it). Will do it today and post here when ready.

**Comment by Utumno** [comment]
Alright - pushed three branches. What I had done originally was pulling https://github.com/Sharlikran/wrye-bash/commit/ce6014cc28eaaf068684fc94d526f5f06b9bca15 and strted working on that. I pushed your original https://github.com/Sharlikran/wrye-bash/commit/ce6014cc28eaaf068684fc94d526f5f06b9bca15 rebased on dev in this adequatelly named (and soon to be deleted) branch:

https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout-ce6014cc28eaaf068684fc94d526f5f06b9bca15-ON-DEV

meanwhile on 16 you pushed https://github.com/Sharlikran/wrye-bash/commit/879f3575f38bf07c595162babfc2424dc1261e6a. Here it is rebased on current dev:

https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout-879f3575f38bf07c595162babfc2424dc1261e6a-ON-DEV

Finally this is my rebase on current dev:

https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout

Compared to https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout-879f3575f38bf07c595162babfc2424dc1261e6a-ON-DEV (your rebase) it misses couple stuff (I will readd asap - namely MelCoed docs, some bsas for fallout 4 and a typo - edit: added apart from melcoed which I will come back to) but also contains some couple things (imports etc) your rebase missed.

Please fetch those on your fork and compare them so you are satisfied and more on mel coed soon

**Comment by Utumno** [comment]
Note on git - those rebases would be a hell haden't I enabled git rerere - which basically remmebers conflict resolutions. So go ahead an enable it it will save you a lot of unneded effort - links in the git tips:

https://github.com/wrye-bash/wrye-bash/wiki/[git]-Tips

**Comment by Utumno** [comment]
And pushed https://github.com/wrye-bash/wrye-bash/commits/151-skyrim-patchers

It's just a subpart of https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout

I left out the cosaves commit (to be amended ASAP) and the melcoed (to be discussed ASAP, I think I now understand) and will leave you some comments soon on couple others.

But this is merge time - yey!

I am eager to merge actually cause yeah - the time has come and this branch contains some important refactoring work also that needs to sit in dev. I mean https://github.com/wrye-bash/wrye-bash/commits/151-skyrim-patchers - the fallout branch I have some thoughts on I will comment on the relevant issue once the dust settles from the skyrim merge.

I have a couple regressions to fix and the BAIN work to finalize but I am eager to hear your comments cause I intend to merge this before Sunday.

I take this opportunity to heartily, publicly, and officially thank you for the incredible effort you put into maintaining this branch. I invite you once again to learn python - you will make a great addition to the team and from the little I know from you you will enjoy it too :)

**Comment by Utumno** [comment]
I have to keep pushing to dev but you don't need to rebase those branches here - just compare them as they are - I will do the final rebase when the merge time comes.

**Comment by Utumno** [comment]
Alright pushed my final https://github.com/wrye-bash/wrye-bash/commits/151-skyrim-patchers and https://github.com/wrye-bash/wrye-bash/commits/151-skyrim-patchers (force necessarily) rebased on dev

Also pushed again https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout-879f3575f38bf07c595162babfc2424dc1261e6a-ON-DEV on current dev so you can compare if you want with your rebase of same commit on current dev

Deleted https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout-ce6014cc28eaaf068684fc94d526f5f06b9bca15-ON-DEV - dead and deprecated

mainly intend to reword the commits to add links to the issues you opened - and maybe have a game constant for typeName dict I commented on here : https://github.com/wrye-bash/wrye-bash/commit/d1152346989ec6a8fbd7b293160e7fed20b08522

Looks ok ?

**Comment by Sharlikran** [comment]
I have feature-sharlikran-valda-fallout rebased again on 151-skyrim-patchers but no idea what is and isn't missing or what fixups I can use and amend to the branches.

**Comment by Utumno** [comment]
> According to gitk feature-sharlikran-valda-fallout-879f3575f38bf07c595162babfc2424dc1261e6a-ON-DEV still points to dev at : faf981b

yep that's current dev :)

Let me pull your branch - done - yep verified on my end - my fallout (bae7bdc96e1614710e6987657f75711815a759ce) and yours (a0c079b84d7c18fb812b5090e7b45d1d86b0490b) differ in those formatting commits, cosaves and record_type_name (which indeed misses from mine, I added it directly to 151 skyrim after I rebased). Will edit a bit your comment commit adding links to the issues you created from that and proceed in merging ASAP

Removing now unneeded https://github.com/wrye-bash/wrye-bash/tree/feature-sharlikran-valda-fallout-879f3575f38bf07c595162babfc2424dc1261e6a-ON-DEV

Going through the diffs took me much more than I expected cause I had to again go through wrapping differences and import differences - it is important you absorb those to your branch.

**Comment by Sharlikran** [comment]
```
>>> import struct
>>> struct.calcsize('=bH') # what is the size in bytes of the structure 'bH' ?
3 # one byte and one H (2 bytes)
>>> struct.calcsize('bH') # what is the size in bytes of the structure 'bH' aligned ?
4 # as above plus a "padding" byte (junk byte to align to 4 bytes)
>>> struct.calcsize('=IHH')
8
>>> struct.calcsize('IHH')
8 # no need to align - already aligned (divisible by four)
```

So this doesn't get lost in a rebase or something I'll add it here.  

However, you don't ever want Wrye to add any padding junk bytes anywhere.  XEdit will know if something is supposed to have 3 bytes but it found 4, which will cause a warning in xEdit.  You wouldn't want that to happen in VMAD since it is very touchy and needs to be formatted exactly however it says in the xEdit code.

I don't know if it will crash Oblivion/Skyrim/Fallout but it doesn't seem like it would be a good idea either.  We decode the records by using small plugins made with the official Bethesda editor so the CK or GECK.  The size should be whatever is specified on the Wiki pages or in the source code.

**Comment by Utumno** [comment]
Where are the record definitions in xEdit source code ?

**Comment by Sharlikran** [comment]
wbDefinitionsFO3.pas
wbDefinitionsFNV.pas
wbDefinitionsFO4.pas
wbDefinitionsTES4.pas
wbDefinitionsTES5.pas

**Comment by Utumno** [comment]
Final version of https://github.com/wrye-bash/wrye-bash/commits/151-skyrim-patchers is up !

please review - I will be merging ASAP :)

Also pushed https://github.com/wrye-bash/wrye-bash/commits/feature-sharlikran-valda-fallout with a fix for cosave names - really took all day (complicated due to enabling disabling saves renames and what not)

**Comment by Utumno** [comment]
Went ahead and pushed :)

I am rebasing your https://github.com/wrye-bash/wrye-bash/commits/feature-sharlikran-valda-fallout - I left you some comments in the current rebase - some of the fallout stuff in there should go to dev I think - I need your feedback on that.

Also if you can update the OP here

Your branches should feel much lighter from now on :)

**Comment by Sharlikran** [comment]
Aside from the Fallout 4 changes mentioned in #268 the LOOT changes should be in dev or people with Wrinkly's newest versions will get errors.

**Comment by Utumno** [comment]
I get the traceback I mentioned with the version in https://github.com/wrye-bash/wrye-bash/commits/feature-sharlikran-valda-fallout - why ?

**Comment by Utumno** [comment]
EDIT: the loot crashing was due to a malformed masterlist I used for testing - still loot must not take Bash down with it - will look into it:

```
Wrye Bash starting
Using Wrye Bash Version 307
Python version: 2.7.10
wxPython version: 2.8.12.1 (msw-unicode)
input encoding: None; output encoding: None; locale: ('en_US', 'cp1253')
filesystem encoding: mbcs 
Searching for game to manage:
bush.py   71 _supportedGames: Detected the following supported games via Windows Registry:
bush.py   73 _supportedGames:  oblivion: C:\GAMES\TESIV\Oblivion
bush.py   73 _supportedGames:  skyrim: C:\GAMES\Skyrim
bush.py  131 _detectGames: Detecting games via the -o argument, bash.ini and relative path:
bush.py  137 _detectGames: Set game mode to oblivion based on sOblivionPath setting in bash.ini:  C:\GAMES\TESIV\Oblivion
bush.py  172 setGame: No preferred game specified.
bush.py  152 __setGame:  Using oblivion game: C:\GAMES\TESIV\Oblivion
testing UAC
mods_metadata.py  219 __init__: Using libbsa API version: 2.0.0
mods_metadata.py  225 __init__: Using LOOT API version: 0.8.1
load_order.py  325 <module>: Using libloadorder API version: 7.6.1
Traceback (most recent call last):
  File "Wrye Bash Launcher.pyw", line 88, in <module>
    bash.main()
  File "bash\bash.py", line 431, in main
    frame = app.Init() # Link.Frame is set here !
  File "bash\basher\__init__.py", line 4194, in Init
    self.InitData(progress)
  File "bash\basher\__init__.py", line 4229, in InitData
    bosh.modInfos.refresh()
  File "bash\bosh\__init__.py", line 3429, in refresh
    self.reloadBashTags()
  File "bash\bosh\__init__.py", line 3607, in reloadBashTags
    mod.reloadBashTags()
  File "bash\bosh\__init__.py", line 2757, in reloadBashTags
    tags = (self.getBashTagsDesc() or set()) | (configHelpers.getBashTags(self.name) or set())
  File "bash\bosh\mods_metadata.py", line 290, in getBashTags
    return self._getTagCache(modName)[0]
  File "bash\bosh\mods_metadata.py", line 303, in _getTagCache
    tags = lootDb.GetModBashTags(modName)
  File "bash\loot.py", line 369, in GetModBashTags
    byref(userlist))
  File "bash\loot.py", line 231, in LootErrorCheck
    raise LootError(result)
bash.loot.LootError: LootError: LOOT_ERROR_NO_TAG_MAP:No Bash Tag map has been previously generated.

```

**Comment by Sharlikran** [comment]
I wouldn't look into it too much.  There are different master list versions with different syntaxes none of which are comparable.  Unless you are using a new version and unless you are pointing to the right git repo then you may have issues.  Even some versions of LOOT require you to delete the old settings files because they aren't comparable.

LOOT release: [Here](https://github.com/loot/loot/releases)
Snapshots: [here](https://www.dropbox.com/sh/scuvwwc6ovzagmd/AAD1TodBAwGQTuV1-4Z2d0sCa?dl=0)

**Comment by Utumno** [comment]
Thanks! I mean if loot crashes Bash should start, this I will make sure (as sure as can be ass now loot craches when accessing Bash tags while it should have crashed when building the db - duh)

**Comment by Sharlikran** [comment]
I don't have time right now but in the next few weeks I'll change the original post to just skyrim records. Because if this is closed but the patchers are not updated I should probably revise it to reflect the current state of Skyrim support.

**Comment by Utumno** [comment]
Ooops - I closed by mistake in the fallout3/nv merge

**Comment by Infernio** [comment]
I just got `Import Factions` to work for all Skyrim-based games 🎉

**Comment by Infernio** [comment]
Was looking at the form list merger and the ideas with porting it to Skyrim, when I noticed this in FO3's `MreFslt`:
https://github.com/wrye-bash/wrye-bash/blob/3426384083bd5d7c61d42730ed7fa9c5629bc2db/Mopy/bash/game/fallout3/records.py#L1562-L1565

That should probably be `self.formIDInList.sort()`, right?

**Comment by Utumno** [comment]
looks like it!

**Comment by Infernio** [comment]
Adding the destructible patcher to Skyrim might make sense in the future: https://www.nexusmods.com/skyrimspecialedition/mods/28291
The mod is still in beta, but seems to work. Destructibles look to be fully functional in Skyrim, just unused.

**Comment by Metallicow** [comment]
ummm....................... no comment
Acquire the real book of destruction. I have one.
Just remember, It multilpies too....!
